src/Lean/Elab/Tactic/BVAckermannize.lean

```diff
@@ -0,0 +1,407 @@
+  let t ← BVTy.ofExpr? (← inferType e)
+  return { x := e, xTy := t}
+
+end Argument
+
+/--
+A function, which packs the expression and the type of the codomain of the function.
+We use the type of the codomain to build an abstracted value for every call of this function.
+-/
+structure Function where
+  /-- The function -/
+  f : Expr
+  /-- The type of the function's codomain -/
+  codomain : BVTy
+ deriving Hashable, BEq, Inhabited
+namespace Function
+
+instance : ToMessageData Function where
+  toMessageData fn := m!"{fn.f} (cod: {fn.codomain})"
+
+/--
+Reify an expression `e` of the kind `f x₁ ... xₙ`, where all the arguments and the return type are
+a `BVTy` into an App.
+-/
+def reifyAp (f : Expr) : OptionT MetaM (Function × Array Argument) := do
+  let xs := f.getAppArgs
+  /- We need at least one argument for this to be a function call we can ackermannize. -/
+  guard <| xs.size > 0
+  let codomain ← BVTy.ofExpr? (← inferType f)
+  let args ← xs.mapM Argument.ofExpr?
+  let fn : Function := { f, codomain }
+  return (fn, args)
+end Function
+/--
+NOTE: is it sensible to hash an array of arguments?
+We may want to use something like a trie to index these.
+Consider swiching to something like `Trie`.
+-/
+abbrev ArgumentList := Array Argument
+/--
+The data stored for an ackermannized call to allow us to build proofs.
+-/
+structure CallVal where
+  /-- The free variable `ack_fx₁...xₙ := (f x₁ x₂ ... xₙ)`. -/
+  fvar : FVarId
+  /-- heqProof : The proof that `ack_fx₁...fxₙ = f x₁ x₂ ... xₙ` (name/fvar = value/expr). -/
+  heqProof : Expr
+deriving Hashable, BEq, Inhabited
+
+namespace CallVal
+
+instance : ToMessageData CallVal where
+  toMessageData cv := m!"{Expr.fvar cv.fvar} ({cv.heqProof})"
+
+end CallVal
+structure State where
+  /--
+  A mapping from a function `f`, to a map of arguments `x₁ ... xₙ`, to the information stored about the call.
+  This is used to generate equations of the form `x₁ = y₁ → x₂ = y₂ → ... → xₙ = yₙ → ack_fx₁...xₙ = ack_fy₁...yₙ on-demand.
+  -/
+  fn2args2call : Std.HashMap Function (Std.HashMap ArgumentList CallVal) := {}
+  /-- A counter for generating fresh names. -/
+  gensymCounter : Nat := 0
+
+def State.init : State where
+
+abbrev AckM := StateRefT State MetaM
+
+namespace AckM
+
+def run (m : AckM α) : MetaM α := m.run' State.init
+
+/-- Generate a fresh name. -/
+def gensym : AckM Name := do
+  modify fun s => { s with gensymCounter := s.gensymCounter + 1 }
+  return Name.mkNum (Name.mkSimple "ackConst") (← get).gensymCounter
+
+def withContext (g : MVarId) (ma : AckM α) : AckM α := g.withContext ma
+
+/-- Get the calls to a function `fn`. -/
+def getCallMap (fn : Function) : AckM (Std.HashMap ArgumentList CallVal) := do
+  return (← get).fn2args2call.getD fn {}
+
+/-- Get the calls to a function `fn`. -/
+def getCallVal? (fn : Function) (args : Array Argument) : AckM (Option CallVal) := do
+  let calls ← getCallMap fn
+  if let .some val := calls.get? args then
```

Comment:
Is this not just `return calls.get? args` ?

Reply:
Fixed, thanks!